internal/beater/tracing.go

```diff
@@ -30,8 +30,7 @@ import (
 	"github.com/elastic/apm-server/internal/model"
 )
 
-func newTracerServer(listener net.Listener, logger *logp.Logger, batchProcessor model.BatchProcessor) (*http.Server, error) {
-	cfg := config.DefaultConfig()
+func newTracerServer(cfg *config.Config, listener net.Listener, logger *logp.Logger, batchProcessor model.BatchProcessor) (*http.Server, error) {
```

Comment:
This was a bit surprising to me, since the modelindexer is passed in. I suppose this is needed for the MaxConcurrentDecoders though, is that right?

This seems like a bit of a problem: it means when self-instrumentation is enabled, there could be 2x the concurrent decoders (1x from external clients, 1x from the server itself). I think this is another reason why the modelindexer should control the rate of input.

(No action required at the moment.)

Reply:
Absolutely right. I thought about using a smaller arbitrary value instead, but like you say I think we can address the problem by encapsulating the semaphore in a different manner and have the model indexer act as the controller.